content/blog/2022/2022-03-31-be-informed-with-lombok.md

```diff
@@ -0,0 +1,552 @@
+   private Gender gender;
+   private String dateOfBirth;
+   private String age;
+   private String socialSecurityNo;
+   private Contact contactDetails;
+   private DriverLicense driverLicense;
+}
+```
+For the project, we have configured a static code analyzer `checkstyle` that runs as a part of the maven `verify` lifecycle.
+In case of the above example (that uses Lombok) the code builds without any issues.
+{{% image alt="settings" src="images/posts/lombok/checkstyle_no_errors.JPG" %}}
+
+In constrast, let's replace the same class with its Delomboked version. After the annotations get replaced with its corresponding constructors, we see issues with the static code analyzer as below.
+{{% image alt="settings" src="images/posts/lombok/checkstyle_with_errors.JPG" %}}
+
+In my experience, I have seen developers use these annotations to escape such violations making it difficult to maintain the code.
+
+### Additional configuration required when using Code Coverage tools
+
+Tools such as **JaCoCo** help create better quality software, as they point out areas of low test coverage in their reports.
+Using Lombok that generates code behind the scenes, greatly affects its code coverage results.
+[Additional configuration](https://reflectoring.io/jacoco/#excluding-code-generated-by-lombok) is required to exclude Lombok generated code.
+
+### @AllArgsConstructor May Introduce Errors When Refactoring
+
+Consider an example class:
+```java
+   @AllArgsConstructor
+   public class Customer {
+   private String id;
+   private String name;
+   private Gender gender;
+   private String dateOfBirth;
+   private String age;
+   private String socialSecurityNo;
+   }
+```
+Let's create an object of Customer class
+```java
+   Customer c = new Customer("C001", "Bryan Rhodes", Gender.MALE, "1986/02/02", "36", "07807789");
+```
+Here, we see that most of the attributes have String as its type. It is easy to mistakenly create an object
+whose params are out of order like this:
+```java
+   Customer c = new Customer("C001", "Bryan Rhodes", Gender.MALE,  "36", "1986/02/02", "07807789");
+```
+If validations are not in place for the attributes, this object might propogate as is in the application.
+Using @Builder here, might avoid such errors.
+
+### @Builder Allows Creation of Invalid Objects
+
+Consider a model as below:
+```java
+    @Builder
+    public class Job {
+        private String id;
+        
+        private JobType jobType;
+    }
+
+    public enum JobType {
+        PLUMBER,
+        BUILDER,
+        CARPENTER
+    }
+```
+For this class, we could construct an object as 
+```java
+    Job job = Job.builder().id("5678").build();
+```
+Although, the code compiles, the object `job` here is in an invalid state because we do not know which `JobType` it belongs to.
+Therefore, along with using the @Builder annotation, it is also important to enforce required attributes to have a value.
+To do this we could consider using **@NonNull** annotation. 
```

Comment:
Can you add a sentence explaining whether it's possible to add custom validations in some way? And maybe add a link to the validation annotations that Lombok supports aside from `NonNull`?

Reply:
There are no other validations in lombok. I have added a reference to the Bean validation API here.